Add wms window

simonseyock opened the pull request:
This resolves #250 

While doing this yesterday I moved some code around. This might be confusing for others, I can undo this. I created a new controller for the layertree and moved the model manipulation from `view/LayerTree.js` into `controller/LayerTreeController.js`.

In my opinion this code belongs into a controller instead of the view.

simonseyock commented:
This is how it looks:

![Peek 2020-09-29 16-47](https://user-images.githubusercontent.com/8100558/94574497-7e656180-0273-11eb-8e63-08d7a2e4a12d.gif)

geographika commented:
@simonseyock - thanks for this. Will be giving a test this morning and get back to you.

simonseyock commented:
I will move the refactoring into a second PR. Is a force push on this branch ok"

geographika commented on app/view/LayerTree.js:
```diff
@@ -27,6 +30,16 @@ Ext.define('CpsiMapview.view.LayerTree', {
+    },
+    tools: [{
+        xtype: 'button',
+        text: 'Add external WMS',
```
Change text to "Add External Map Layer"

geographika commented:
Thanks @simonseyock - I have this integrated and working in one of full applications. 

A few questions:

- How do we configure the button text in the form to be in English? Does this require updates to BasiGX?
- How is the WMS URL configured (currently set to https://ows.terrestris.de/osm/service)? Is it possible to have a dropdown box with some preset servers plus allow free text entry?
- Is it possible to add in the layer "Refresh" and "Opacity" layer plugins to the added layers?

![image](https://user-images.githubusercontent.com/490840/94672635-d2288700-0315-11eb-987b-cb61fcde0e8a.png)

For my own reference (but please clarify if incorrect), to add to an application based on cpsi-mapview:

- Add a reference to `CpsiMapview.view.addWms.AddWmsWindow` to the application Main view 
- Add a new `xtype: 'cmv_add_wms_window'` to the main view. This is the form that allows a user to add a WMS. It is hidden by default, and reused when a user clicks on the button in the layer tree

When the form is created it adds a listener to the button click event to be displayed. 
When the first layer is added a new group layer is added to the map (and hence the layer tree).

simonseyock commented:
I will look into your comments. Is a force push on this branch ok?

simonseyock commented on app/view/addWms/AddWmsWindow.js:
```diff
@@ -0,0 +1,20 @@
+Ext.define('CpsiMapview.view.addWms.AddWmsWindow', {
+    xtype: 'cmv_add_wms_window',
+
+    extend: 'Ext.window.Window',
```
Yes this is no problem.

simonseyock commented on app/controller/LayerTreeController.js:
```diff
@@ -0,0 +1,205 @@
+    extend: 'Ext.app.ViewController',
+
+    alias: 'controller.cmv_layertree',
+
```
I add these changes to refactor PR

simonseyock commented:
> How do we configure the button text in the form to be in English? Does this require updates to BasiGX?
The translations can be added by changing the viewModel of the `view.addWms.AddWmsForm`.

I added some translations.

> How is the WMS URL configured (currently set to https://ows.terrestris.de/osm/service)? Is it possible to have a dropdown box with some preset servers plus allow free text entry?

This can be done easily with the property `wmsBaseUrls` on the `view.addWms.AddWmsForm` like this:
```
    wmsBaseUrls: [
        'https://ows.terrestris.de/osm/service',
        'url2'
    ],
```
It automatically shows a combo box and text input is still allowed.

> Is it possible to add in the layer "Refresh" and "Opacity" layer plugins to the added layers?

I added the options to the layers.

simonseyock commented:
> For my own reference (but please clarify if incorrect), to add to an application based on cpsi-mapview:
>
>    Add a reference to CpsiMapview.view.addWms.AddWmsWindow to the application Main view
>    Add a new xtype: 'cmv_add_wms_window' to the main view. This is the form that allows a user to add a WMS. It is hidden by default, and reused when a user clicks on the button in the layer tree
>
> When the form is created it adds a listener to the button click event to be displayed.
> When the first layer is added a new group layer is added to the map (and hence the layer tree).

This is correct. To change the name of the layerGroup the property `layerGroupName` of the `view.addWms.AddWmsForm` can be changed.

geographika commented:
> 
> 
> I will look into your comments. Is a force push on this branch ok?

Yes fine with me.

simonseyock commented:
I added the requested changes. This branch is now rebased on layer-tree-refactoring.

geographika commented:
Thanks @simonseyock. I'm happy to merge this. 

For reference - this is how I'm adding and configuring in an application:

```js
         items: [
...
// layer tree declaration
...
        {
            xtype: 'cmv_add_wms_window',
            title: 'Add External WMS Layer',
            items: [
                {
                    xtype: 'cmv_add_wms_form',
                    wmsBaseUrls: [
                        'https://ows.terrestris.de/osm/service',
                        'url2'
                    ]
                }
            ]
          }
...
        ]
```

One thing to watch out for is that the form has to be added to the viewport's items **after** the layertree component declaration (I'm adding it outside the layertree otherwise it is constrained to the layertree panel when moving the window). 
Otherwise, the following line in `setupListener` returns an empty list as the layertree component is not created:

`var layerTrees = Ext.ComponentQuery.query('cmv_layertree');`

chrismayer commented on app/controller/addWms/AddWmsFormController.js:
```diff
@@ -0,0 +1,32 @@
+
+    alias: 'controller.cmv_add_wms_form',
+
+    onWmsAdd: function (olLayer) {
```
Please add some API docs with a short description here.

chrismayer commented on app/controller/addWms/AddWmsWindowController.js:
```diff
@@ -0,0 +1,19 @@
+        this.setupListener();
+    },
+
+    setupListener: function () {
```
Please add some API docs with a short description here.

chrismayer commented on app/view/LayerTree.js:
```diff
@@ -28,6 +28,16 @@ Ext.define('CpsiMapview.view.LayerTree', {
+    header: {
+        dock: 'bottom'
+    },
+    tools: [{
```
Something like 

```
bbar: [{
  text: 'Add External Map Layer',
...
}]
```
should do the same and would make 

```
 header: {
        dock: 'bottom'
    },
```
obsolete.

Just for information, no need to change.

chrismayer commented on app/view/addWms/AddWmsForm.js:
```diff
@@ -0,0 +1,44 @@
+            wmsVersionContainerFieldLabel: 'Version',
+            availableLayesFieldSetTitle: 'Available layers',
+            resetBtnText: 'Reset',
+            requestLayersBtnText: 'Request available Layers',
```
```suggestion
            requestLayersBtnText: 'Request available layers',
```

chrismayer commented on app/view/addWms/AddWmsForm.js:
```diff
@@ -0,0 +1,44 @@
+            errorRequestFailed: 'The given URL could not be requested.',
+            errorCouldntParseResponse: 'The received response could not be parsed successfully.',
+            msgRequestTimedOut: 'The request was not answered in time and was aborted.',
+            msgServiceException: 'A OGC ServiceException has been occurred',
```
```suggestion
            msgServiceException: 'An OGC ServiceException has been occurred',
```

marcjansen replied:
```suggestion
            msgServiceException: 'An OGC ServiceException has occurred',
```

chrismayer commented on app/view/addWms/AddWmsForm.js:
```diff
@@ -0,0 +1,44 @@
+            msgServiceUnavailable: 'The server is currently not available (too many or request or maintaince mode).',
+            msgGatewayTimeOut: 'The server acted as a gateway and the original target did not respond in time',
+            msgClientError: 'An unspecified client error occurred.',
+            msgServerError: 'En unspecified server error occured.',
```
```suggestion
            msgServerError: 'An unspecified server error occurred.',
```

chrismayer commented on app/view/main/Main.js:
```diff
@@ -44,5 +45,7 @@ Ext.define('CpsiMapview.view.main.Main', {
     }, {
         xtype: 'cmv_header'
+    }, {
+        xtype: 'cmv_add_wms_window'
```
How about instantiating the `cmv_add_wms_window` in the handler of the button instead of hanging it in the main view?

chrismayer reviewed (commented):
Very cool @simonseyock. I added some minor issues you might want to have a look at before merging. TIA!

simonseyock commented:
I added the suggestions. The window is now created inside the LayerTreeController like this:
```
            this.addWmsWindow = Ext.create('CpsiMapview.view.window.MinimizableWindow', {
                items: [
                    {
                        xtype: 'cmv_add_wms_form'
                    }
                ],
            });
            this.addWmsWindow.show();
```
This way it does not need to be part of the main view.

geographika commented:
@simonseyock - there is a conflict to fix due to updates to the LayerTreeController. After those are fixed I think this is fine to merge - thoughts @chrismayer ?

chrismayer commented:
Yes, this is good to go (after resolving).

simonseyock commented:
@geographika I resolved the merge conflicts
